src/Computation/Expression.cs

```diff
@@ -2,308 +2,500 @@
+                : SymbolicProposition.Create((IPersistentSpace) space, _value, _constraints ?? Array.Empty<IValue>());
         }
 
         public IExpression Add(IExpression expression)
         {
-            return Create(expression, (l, r) => l.AsUnsigned().Add(r.AsUnsigned()));
+            return Binary(expression,
+                (l, r) => new Add(l, r),
+                (l, r) => ConstantUnsigned.Create(Size, (BigInteger) l.AsUnsigned() + r.AsUnsigned()));
         }
 
         public IExpression And(IExpression expression)
         {
-            return Create(expression, (l, r) => l.AsBitwise().And(r.AsBitwise()));
+            return Binary(expression,
+                (l, r) => new And(l, r),
+                (l, r) => ConstantUnsigned.Create(Size, (BigInteger) l.AsUnsigned() & r.AsUnsigned()));
         }
 
         public IExpression ArithmeticShiftRight(IExpression expression)
         {
-            return Create(expression, (l, r) => l.AsSigned().ArithmeticShiftRight(r.AsUnsigned()));
+            return Binary(expression,
+                (l, r) => new ArithmeticShiftRight(l, r),
+                (l, r) => ConstantSigned.Create(Size, (BigInteger) l.AsSigned() >> (int) r.AsUnsigned()));
         }
 
         public IExpression Equal(IExpression expression)
         {
-            return Create(expression, (l, r) => l.AsBitwise().Equal(r.AsBitwise()));
+            return Binary(expression,
+                (l, r) => new Equal(l, r),
+                (l, r) => new ConstantBool((BigInteger) l.AsUnsigned() == r.AsUnsigned()));
         }
 
         public IExpression FloatAdd(IExpression expression)
         {
-            return Create(expression, (l, r) => l.AsFloat().Add(r.AsFloat()));
+            return Binary(expression,
+                (l, r) => new FloatAdd(l, r),
+                (l, r) => new ConstantSingle(l + r),
+                (l, r) => new ConstantDouble(l + r));
         }
 
         public IExpression FloatCeiling()
         {
-            return Create(c => c.AsFloat().Ceiling());
+            return Unary(
+                v => new FloatCeiling(v),
+                v => new ConstantSingle(MathF.Ceiling(v)),
+                v => new ConstantDouble(Math.Ceiling(v)));
         }
 
         public IExpression FloatConvert(Bits size)
         {
-            return Create(c => c.AsFloat().Convert(size));
+            return Unary(
+                v => v is IRealValue r
+                    ? new RealConvert(size, r)
+                    : new FloatConvert(size, v),
+                v => (uint) size switch
+                {
+                    32U => _value,
+                    64U => new ConstantDouble(v),
+                    _ => new FloatConvert(size, _value)
+                },
+                v => (uint) size switch
+                {
+                    32U => new ConstantSingle((float) v),
+                    64U => _value,
+                    _ => new FloatConvert(size, _value)
+                });
```

Comment:
This float thing is the one thing that I would say becomes a bit worse, but I think we just have to eat this.